File: software/mobile/ios/SVCMobile/Tests/RideDashboardTests.swift
import XCTest
@testable import SVCMobile

final class RideDashboardTests: XCTestCase {
    private var profile: VehiclePerformanceProfile {
        VehiclePerformanceCatalog.load(bundle: Bundle(for: Self.self))
            .resolve(profileId: "bmw-r1200gs-k25-2007")
    }

    func testTachometerBoundariesAreExact() {
        let expected: [(Double, TachometerZone)] = [
            (0, .normal),
            (6_999, .normal),
            (7_000, .warning),
            (7_799, .warning),
            (7_800, .red),
            (9_000, .red)
        ]

        for (rpm, zone) in expected {
            XCTAssertEqual(
                TachometerZoneResolver.zone(rpm: rpm, profile: profile),
                zone
            )
        }
        XCTAssertNil(profile.revLimiterRpm)
    }

    func testTachometerHysteresisAvoidsBoundaryOscillation() {
        XCTAssertEqual(
            TachometerZoneResolver.zoneWithHysteresis(
                rpm: 7_750,
                profile: profile,
                previous: .red
            ),
            .red
        )
        XCTAssertEqual(
            TachometerZoneResolver.zoneWithHysteresis(
                rpm: 6_950,
                profile: profile,
                previous: .warning
            ),
            .warning
        )
    }

    func testAllGearPresentationStatesAreExplicit() {
        XCTAssertEqual(
            RideGear.allCases.map(\.displayValue),
            ["N", "1", "2", "3", "4", "5", "6", "BETWEEN", "—"]
        )
    }

    func testStaleInvalidUnavailableAndDegradedStates() {
        XCTAssertNil(
            RideValue(value: 1_200.0, unit: "rpm", state: .stale).displayValue
        )
        XCTAssertNil(
            RideValue(value: 1_200.0, unit: "rpm", state: .invalid).displayValue
        )
        XCTAssertNil(
            RideValue(value: 1_200.0, unit: "rpm", state: .unavailable).displayValue
        )
        XCTAssertEqual(
            RideValue(value: 1_200.0, unit: "rpm", state: .degraded).displayValue,
            1_200
        )
    }

    func testLeanExtremaTrackBothDirectionsAndResetOnlyAtZero() {
        var extrema = LeanExtrema()
        extrema.observe(-34)
        extrema.observe(42)
        extrema.observe(-12)
        XCTAssertEqual(extrema.maximumLeftDegrees, 34)
        XCTAssertEqual(extrema.maximumRightDegrees, 42)

        XCTAssertFalse(
            extrema.resetIfStationary(
                speed: RideValue(value: 2.0, unit: "km/h", state: .valid)
            )
        )
        XCTAssertEqual(extrema.maximumLeftDegrees, 34)
        XCTAssertTrue(
            extrema.resetIfStationary(
                speed: RideValue(value: 0.0, unit: "km/h", state: .valid)
            )
        )
        XCTAssertEqual(extrema, LeanExtrema())
    }

    func testCalibrationRequiresDemoAndConfirmedStop() {
        let stopped = RideValue(value: 0.0, unit: "km/h", state: .valid)
        let moving = RideValue(value: 1.0, unit: "km/h", state: .valid)
        let unavailable = RideValue<Double>(
            value: nil,
            unit: "km/h",
            state: .unavailable
        )

        XCTAssertTrue(
            MotorcycleCalibrationPolicy.canCalibrate(
                speed: stopped,
                demoMode: true
            )
        )
        XCTAssertFalse(
            MotorcycleCalibrationPolicy.canCalibrate(
                speed: stopped,
                demoMode: false
            )
        )
        XCTAssertFalse(
            MotorcycleCalibrationPolicy.canCalibrate(
                speed: moving,
                demoMode: true
            )
        )
        XCTAssertFalse(
            MotorcycleCalibrationPolicy.canCalibrate(
                speed: unavailable,
                demoMode: true
            )
        )
    }

    func testAutomaticThemeHysteresisAndUnavailableLight() {
        let low = RideValue(value: 200.0, unit: "lux", state: .valid)
        let middle = RideValue(value: 450.0, unit: "lux", state: .valid)
        let high = RideValue(value: 700.0, unit: "lux", state: .valid)
        let unavailable = RideValue<Double>(
            value: nil,
            unit: "lux",
            state: .unavailable
        )

        XCTAssertEqual(
            RideThemeResolver.resolve(
                mode: .automatic,
                ambientLight: low,
                previous: .day
            ),
            .night
        )
        XCTAssertEqual(
            RideThemeResolver.resolve(
                mode: .automatic,
                ambientLight: middle,
                previous: .night
            ),
            .night
        )
        XCTAssertEqual(
            RideThemeResolver.resolve(
                mode: .automatic,
                ambientLight: high,
                previous: .night
            ),
            .day
        )
        XCTAssertEqual(
            RideThemeResolver.resolve(
                mode: .automatic,
                ambientLight: unavailable,
                previous: .day
            ),
            .day
        )
    }

    func testProfileSwitchFallsBackSafely() {
        let catalog = VehiclePerformanceCatalog.load(bundle: Bundle(for: Self.self))
        XCTAssertEqual(
            catalog.resolve(profileId: "bmw-r1200gs-k25-2007").redZoneStartRpm,
            7_800
        )
        XCTAssertNil(
            catalog.resolve(profileId: "generic-motorcycle").redZoneStartRpm
        )
        XCTAssertEqual(
            catalog.resolve(profileId: "missing-profile").id,
            "generic-motorcycle"
        )
    }

    func testMockDashboardDoesNotInventGearAndKeepsWarning() {
        let snapshot = MockDeviceRepository(
            bundle: Bundle(for: Self.self)
        ).currentSnapshot()
        let dashboard = RideDashboardState.build(
            telemetry: snapshot.telemetry,
            connectionState: snapshot.connectionState,
            isConnecting: false,
            profile: profile
        )

        XCTAssertEqual(dashboard.gear, .unavailable)
        XCTAssertEqual(dashboard.leanAngle.state, .degraded)
        XCTAssertTrue(dashboard.alerts.contains { $0.id == "sd_card_missing" })
    }

    func testReduceMotionUsesZeroDuration() {
        XCTAssertEqual(RideMotionPolicy.duration(reduceMotion: true), 0)
        XCTAssertEqual(
            RideMotionPolicy.duration(reduceMotion: false),
            RideDesignTokens.standardAnimation
        )
    }
}

File: software/mobile/android/core-model/src/test/kotlin/com/avlyubimov/svc/core/model/RideDashboardTests.kt
package com.avlyubimov.svc.core.model

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class RideDashboardTests {
    private val profile = VehiclePerformanceProfile(
        schemaVersion = 1,
        id = "bmw-r1200gs-k25-2007",
        manufacturer = "BMW",
        model = "R1200GS",
        generation = "K25",
        yearFrom = 2007,
        yearTo = 2007,
        engineName = "1170 cc two-cylinder boxer",
        engineDisplacementCc = 1170.0,
        maximumTorqueNm = 115.0,
        nominalPowerKw = 74.0,
        gearboxGears = 6,
        fuelCapacityLiters = 20.0,
        fuelReserveLiters = 4.0,
        iceWarningTemperatureCelsius = 3.0,
        idleRpm = 1150,
        idleToleranceRpm = 50,
        torquePeakRpm = 5500,
        powerPeakRpm = 7000,
        tachometerScaleMinRpm = 0,
        tachometerScaleMaxRpm = 9000,
        warningStartRpm = 7000,
        redZoneStartRpm = 7800,
        revLimiterRpm = null,
        source = "test",
        reference = "test",
        confidence = "test",
    )

    @Test
    fun tachometerBoundariesAreExact() {
        val expected = mapOf(
            0.0 to TachometerZone.NORMAL,
            6999.0 to TachometerZone.NORMAL,
            7000.0 to TachometerZone.WARNING,
            7799.0 to TachometerZone.WARNING,
            7800.0 to TachometerZone.RED,
            9000.0 to TachometerZone.RED,
        )
        expected.forEach { (rpm, zone) ->
            assertEquals(zone, TachometerZoneResolver.zone(rpm, profile))
        }
        assertNull(profile.revLimiterRpm)
    }

    @Test
    fun tachometerUsesBoundaryHysteresis() {
        assertEquals(
            TachometerZone.RED,
            TachometerZoneResolver.zoneWithHysteresis(
                rpm = 7750.0,
                profile = profile,
                previous = TachometerZone.RED,
            ),
        )
        assertEquals(
            TachometerZone.WARNING,
            TachometerZoneResolver.zoneWithHysteresis(
                rpm = 6950.0,
                profile = profile,
                previous = TachometerZone.WARNING,
            ),
        )
    }

    @Test
    fun allGearPresentationStatesAreExplicit() {
        assertEquals(
            listOf("N", "1", "2", "3", "4", "5", "6", "BETWEEN", "—"),
            RideGear.entries.map(RideGear::displayValue),
        )
    }

    @Test
    fun staleInvalidAndUnavailableNeverDisplayValues() {
        assertNull(RideValue(1200.0, "rpm", RideDataState.STALE).displayValue)
        assertNull(RideValue(1200.0, "rpm", RideDataState.INVALID).displayValue)
        assertNull(RideValue(1200.0, "rpm", RideDataState.UNAVAILABLE).displayValue)
        assertEquals(
            1200.0,
            RideValue(1200.0, "rpm", RideDataState.DEGRADED).displayValue,
        )
    }

    @Test
    fun leanExtremaTrackBothSidesAndResetOnlyAtZero() {
        val extrema = LeanExtrema()
            .observe(-34.0)
            .observe(42.0)
            .observe(-12.0)
        assertEquals(34.0, extrema.maximumLeftDegrees)
        assertEquals(42.0, extrema.maximumRightDegrees)
        assertEquals(
            extrema,
            extrema.resetIfStationary(
                RideValue(2.0, "km/h", RideDataState.VALID),
            ),
        )
        assertEquals(
            LeanExtrema(),
            extrema.resetIfStationary(
                RideValue(0.0, "km/h", RideDataState.VALID),
            ),
        )
    }

    @Test
    fun calibrationRequiresDemoAndConfirmedStationarySpeed() {
        val stopped = RideValue(0.0, "km/h", RideDataState.VALID)
        val moving = RideValue(1.0, "km/h", RideDataState.VALID)
        val unavailable = RideValue<Double>(
            null,
            "km/h",
            RideDataState.UNAVAILABLE,
        )
        assertTrue(MotorcycleCalibrationPolicy.canCalibrate(stopped, demoMode = true))
        assertFalse(MotorcycleCalibrationPolicy.canCalibrate(stopped, demoMode = false))
        assertFalse(MotorcycleCalibrationPolicy.canCalibrate(moving, demoMode = true))
        assertFalse(
            MotorcycleCalibrationPolicy.canCalibrate(unavailable, demoMode = true),
        )
    }

    @Test
    fun automaticThemeUsesHysteresisAndRetainsUnknownState() {
        val low = RideValue(200.0, "lux", RideDataState.VALID)
        val middle = RideValue(450.0, "lux", RideDataState.VALID)
        val high = RideValue(700.0, "lux", RideDataState.VALID)
        val unavailable = RideValue<Double>(
            null,
            "lux",
            RideDataState.UNAVAILABLE,
        )

        assertEquals(
            RideResolvedTheme.NIGHT,
            RideThemeResolver.resolve(
                RideThemeMode.AUTOMATIC,
                low,
                RideResolvedTheme.DAY,
            ),
        )
        assertEquals(
            RideResolvedTheme.NIGHT,
            RideThemeResolver.resolve(
                RideThemeMode.AUTOMATIC,
                middle,
                RideResolvedTheme.NIGHT,
            ),
        )
        assertEquals(
            RideResolvedTheme.DAY,
            RideThemeResolver.resolve(
                RideThemeMode.AUTOMATIC,
                high,
                RideResolvedTheme.NIGHT,
            ),
        )
        assertEquals(
            RideResolvedTheme.DAY,
            RideThemeResolver.resolve(
                RideThemeMode.AUTOMATIC,
                unavailable,
                RideResolvedTheme.DAY,
            ),
        )
    }

    @Test
    fun reduceMotionDisablesDashboardAnimationDuration() {
        assertEquals(0, RideMotionPolicy.durationMillis(reduceMotion = true))
        assertEquals(180, RideMotionPolicy.durationMillis(reduceMotion = false))
        assertEquals(
            300,
            RideMotionPolicy.durationMillis(
                reduceMotion = false,
                standardDurationMillis = 300,
            ),
        )
    }
}
